message_ix_models/tools/iea/weo.py

```diff
@@ -0,0 +1,672 @@
+"""
+Code for handling IEA WEO data
+"""
+
+import numpy as np
+import pandas as pd
+
+from message_ix_models.util import package_data_path
+
+
+def get_weo_data():
+    """
+    Read in raw WEO investment/capital costs and O&M costs data
+    (for all technologies and for STEPS scenario only).
+    Convert to long format
+
+    Returns DataFrame of processed data
+    """
+
+    # Read in raw data file
+    file_path = package_data_path(
+        "iea", "WEO_2022_PG_Assumptions_STEPSandNZE_Scenario.xlsb"
+    )
+
+    # Dict of all of the technologies,
+    # their respective sheet in the Excel file,
+    # and the start row
+    tech_rows = {
+        "steam_coal_subcritical": ["Coal", 5],
+        "steam_coal_supercritical": ["Coal", 15],
+        "steam_coal_ultrasupercritical": ["Coal", 25],
+        "igcc": ["Coal", 35],
+        "ccgt": ["Gas", 5],
+        "gas_turbine": ["Gas", 15],
+        "ccgt_chp": ["Gas", 25],
+        "fuel_cell": ["Gas", 35],
+        "pulverized_coal_ccs": ["Fossil fuels equipped with CCUS", 5],
+        "igcc_ccs": ["Fossil fuels equipped with CCUS", 15],
+        "ccgt_ccs": ["Fossil fuels equipped with CCUS", 25],
+        "nuclear": ["Nuclear", 5],
+        "solarpv_large": ["Renewables", 5],
+        "solarpv_buildings": ["Renewables", 15],
+        "wind_onshore": ["Renewables", 25],
+        "wind_offshore": ["Renewables", 35],
+        "hydropower_large": ["Renewables", 45],
+        "hydropower_small": ["Renewables", 55],
+        "bioenergy_large": ["Renewables", 65],
+        "bioenergy_cofiring": ["Renewables", 75],
+        "bioenergy_medium_chp": ["Renewables", 85],
+        "bioenergy_ccus": ["Renewables", 95],
+        "csp": ["Renewables", 105],
+        "geothermal": ["Renewables", 115],
+        "marine": ["Renewables", 125],
+    }
+
+    # Specify cost types to read in and the required columns
+    cost_cols = {"capital_costs": "A,B:D", "annual_om_costs": "A,F:H"}
+
+    # Loop through each technology and cost type
+    # Read in data and convert to long format
+    dfs_cost = []
+    for tech_key in tech_rows:
```

Comment:
Use [`itertools.product()`](https://docs.python.org/3/library/itertools.html#itertools.product) instead of nesting loops.